// fair/lib/build/base.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t;
    return { next: verb(0), "throw": verb(1), "return": verb(2) };
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var path = require("path");
var BaseBuilder = (function () {
    function BaseBuilder(hay) {
        this.hay = hay;
        this.WATCH_INITIATED = false;
        this.progressBar = this.hay.reporter.createProgressBar();
    }
    BaseBuilder.prototype.loadFile = function (fileName) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.hay.fileSystem
                        .readFile(path.resolve(this.config.directory, fileName))
                        .then(function (contents) {
                        _this.progressBar.tick("loaded " + fileName);
                        return {
                            fileName: fileName,
                            contents: contents
                        };
                    })];
            });
        });
    };
    BaseBuilder.prototype.loadFiles = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var files;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.hay.fileSystem.readDir(this.config.directory)];
                    case 1:
                        files = _a.sent();
                        files = files.filter(this.checkExtensions(this.config.fileExtensions));
                        this.progressBar.setLength(files.length);
                        this.progressBar.setCategory("load " + this.config.name);
                        this.progressBar.start();
                        return [2 /*return*/, Promise.all(files.map(function (fileName) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                                return [2 /*return*/, this.loadFile(fileName)];
                            }); }); }))];
                }
            });
        });
    };
    BaseBuilder.prototype.setConfig = function (config) {
        this.config = config;
    };
    BaseBuilder.prototype.checkExtensions = function (extensions) {
        var _this = this;
        return function (file) { return extensions.includes(_this.hay.fileSystem.getFileExtension(file)); };
    };
    return BaseBuilder;
}());
exports.BaseBuilder = BaseBuilder;
